tox.ini

```diff
@@ -59,7 +59,7 @@ commands_pre = coverage html --fail-under=0
 commands = coverage report
 
 [testenv:lint]
-deps = pre-commit~=2.9.2
```

Comment:
I'm not yet convinced this should be pinned; I would anticipate that pinning the hook versions, not pre-commit, is acceptable here.

Reply:
Yeah, I'm not sure. I almost wrote `<4` as the version bound. Do we want to go with that for now?

We sometimes need a lower bound (which can be implicit; we'll assume a relatively recent one if unspecified) to make sure newer features work, like `types_or`. I think we want an explicit upper bound in case there are new format changes (v3 was, I think, when the old config format was removed?) or other breaking changes which I can't foresee.

Give me an indication of which you'd prefer: pinned or `<4`. I can't think of alternatives, but if there is one that's compelling, that would be cool too.

Reply:
I think my gut here is "linting isn't a default environment, and pre-commit.ci is enabled for this repo".

Therefore I'd set `recreate = true` in the tox environment and then remove the version specifier completely. Tox will always recreate the environment and install the latest version, which will almost certainly identical behavior to what we'll be getting from pre-commit.ci anyway.

Reply:
I just circled back to this and made the change. I'm sort of on the fence now that I look at it -- maybe we should just remove `tox r -e lint`... but for now unpinned with `recreate=true` will do the trick!